`patchState` TypeScript v5.4 compilation error

### Which @ngrx/* package(s) are the source of the bug?

signals

### Minimal reproduction of the bug/regression with instructions

https://stackblitz.com/edit/stackblitz-starters-weemkm?file=src%2Fmain.ts

If you chain multiple updaters in `patchState`, with one being an updater from the `@ngrx/signals/entities` package, you now get a compilation error when upgrading to TypeScript v5.4.3:

```ts
type CoreState = { loading: boolean };
type Entity = { id: string };

const Store = signalStore(
  withState<CoreState>({ loading: true }),
  withEntities<Entity>(),
  withMethods((store) => {
    const setLoading = () => {
      const newState: CoreState = { loading: true };
      patchState(store, newState, removeAllEntities());
    };

    return {};
  })
);
```

```
✘ [ERROR] TS2345: Argument of type 'CoreState' is not assignable to parameter of type 'PartialStateUpdater<EntityState<any>> | Partial<EntityState<any>>'. [plugin angular-compiler]

    src/main.ts:20:24:
      20 │       patchState(store, newState, removeAllEntities());
```

This previously worked fine in TypeScript v5.3.

### Expected behavior

To compile (and work) fine.

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

NgRx Signals: v17.1.1
Angular: v17.3.0
TypeScript: v5.4.3

### Other information

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

--- a/modules/signals/spec/types/patch-state.types.spec.ts
+++ b/modules/signals/spec/types/patch-state.types.spec.ts
@@ -1,0 +1,72 @@
+import { expecter } from 'ts-snippet';
+import { compilerOptions } from './helpers';
+
+describe('patchState', () => {
+  const expectSnippet = expecter(
+    (code) => `
+        import {
+          PartialStateUpdater,
+          patchState,
+          signalState,
+        } from '@ngrx/signals';
+
+        const state = signalState({ count: 1, foo: 'bar' });
+
+        function increment(): PartialStateUpdater<{ count: number }> {
+          return ({ count }) => ({ count: count + 1 });
+        }
+
+        function addNumber(num: number): PartialStateUpdater<{
+          numbers: number[];
+        }> {
+          return ({ numbers }) => ({ numbers: [...numbers, num] });
+        }
+
+        ${code}
+      `,
+    compilerOptions()
+  );
+
+  it('infers the state type from StateSignal', () => {
+    expectSnippet('patchState(state, increment())').toSucceed();
+    expectSnippet("patchState(state, { foo: 'baz' })").toSucceed();
+    expectSnippet("patchState(state, { foo: 'baz' }, increment())").toSucceed();
+    expectSnippet("patchState(state, increment(), { foo: 'baz' })").toSucceed();
+  });
+
+  it('fails with wrong partial state object', () => {
+    expectSnippet('patchState(state, { x: 1 })').toFail(
+      /'x' does not exist in type 'Partial<{ count: number; foo: string; }>/
+    );
+    expectSnippet("patchState(state, { foo: 'baz' }, { x: 1 })").toFail(
+      /'x' does not exist in type 'Partial<{ count: number; foo: string; }>/
+    );
+    expectSnippet('patchState(state, { x: 1 }, { count: 0 })').toFail(
+      /'x' does not exist in type 'Partial<{ count: number; foo: string; }>/
+    );
+    expectSnippet('patchState(state, increment(), { x: 1 })').toFail(
+      /'x' does not exist in type 'Partial<{ count: number; foo: string; }>/
+    );
+    expectSnippet('patchState(state, { x: 1 }, increment())').toFail(
+      /'x' does not exist in type 'Partial<{ count: number; foo: string; }>/
+    );
+  });
+
+  it('fails with wrong partial state updater', () => {
+    expectSnippet('patchState(state, addNumber(10))').toFail(
+      /Property 'numbers' is missing in type '{ count: number; foo: string; }'/
+    );
+    expectSnippet('patchState(state, { count: 10 }, addNumber(10))').toFail(
+      /Property 'numbers' is missing in type '{ count: number; foo: string; }'/
+    );
+    expectSnippet('patchState(state, addNumber(10), { count: 10 })').toFail(
+      /Property 'numbers' is missing in type '{ count: number; foo: string; }'/
+    );
+    expectSnippet('patchState(state, increment(), addNumber(10))').toFail(
+      /Property 'numbers' is missing in type '{ count: number; foo: string; }'/
+    );
+    expectSnippet('patchState(state, addNumber(10), increment())').toFail(
+      /Property 'numbers' is missing in type '{ count: number; foo: string; }'/
+    );
+  });
+});

--- a/modules/signals/src/patch-state.ts
+++ b/modules/signals/src/patch-state.ts
@@ -1,4 +1,5 @@
 import { STATE_SIGNAL, StateSignal } from './state-signal';
+import { Prettify } from './ts-helpers';
 
 export type PartialStateUpdater<State extends object> = (
   state: State
@@ -6,7 +7,9 @@ export type PartialStateUpdater<State extends object> = (
 
 export function patchState<State extends object>(
   stateSignal: StateSignal<State>,
-  ...updaters: Array<Partial<State & {}> | PartialStateUpdater<State & {}>>
+  ...updaters: Array<
+    Partial<Prettify<State>> | PartialStateUpdater<Prettify<State>>
+  >
 ): void {
   stateSignal[STATE_SIGNAL].update((currentState) =>
     updaters.reduce(

